components/Rect.tsx

```diff
@@ -1,21 +1,29 @@
 import { useContext2d } from "./Canvas.tsx";
-import { useEffect } from "preact/hooks";
+
+export type RectProps = {
+  children?: null | preact.JSX.Element;
+  fillColor: string | CanvasGradient | CanvasPattern;
+  height?: number;
+  width?: number;
+  x?: number;
+  y?: number;
+};
 
 export function Rect({
+  children = null,
   fillColor,
-  x,
-  y,
-}: {
-  fillColor: string;
-  x: number;
-  y: number;
-}) {
+  height: controlledHeight = -1,
+  width: controlledWidth = -1,
+  x = 0,
+  y = 0,
+}: RectProps) {
   const context2d = useContext2d();
-  useEffect(() => {
-    console.log("Rect", x, y, fillColor);
-    context2d.fillStyle = fillColor;
-    context2d.fillRect(x, y, context2d.canvas.width, context2d.canvas.height);
-  }, [context2d, fillColor, x, y]);
+  const width =
+    controlledWidth === -1 ? context2d.canvas.width : controlledWidth;
+  const height =
+    controlledHeight === -1 ? context2d.canvas.height : controlledHeight;
+  context2d.fillStyle = fillColor;
+  context2d.fillRect(x, y, width, height);
```

Comment:
instead of rendering in a useEffect, we render during the rendering phase, which fixes the random white frames we were seeing last night 🎉